ompi/mca/osc/ucx/osc_ucx_component.c

```diff
@@ -0,0 +1,565 @@
+            }
+        }
+
+        ucp_worker_release_address(mca_osc_ucx_component.ucp_worker, my_addr);
+        my_addr = NULL;
+        free(recv_buf);
+        recv_buf = NULL;
+    }
+
+    /* initialize UCP win_info array */
+    module->win_info_array = calloc(comm_size, sizeof(ompi_osc_ucx_win_info_t));
+    if (module->win_info_array == NULL) {
+        goto error;
+    }
+
+    ret = mem_map_and_exchange_win_info(base, size, &(module->memh), module,
+                                        module->win_info_array, flavor);
+    if (ret != OMPI_SUCCESS) {
+        goto error;
+    }
+
+    /* initialize UCP state_info array */
+    module->state_info_array = calloc(comm_size, sizeof(ompi_osc_ucx_win_info_t));
+    if (module->state_info_array == NULL) {
+        goto error;
+    }
+
+    state_base = (void *)&(module->state);
+    ret = mem_map_and_exchange_win_info(&state_base, sizeof(ompi_osc_ucx_state_t),
+                                        &(module->state_memh), module,
+                                        module->state_info_array,
+                                        MPI_WIN_FLAVOR_CREATE);
+    if (ret != OMPI_SUCCESS) {
+        goto error;
+    }
+
+    module->state.lock = TARGET_LOCK_UNLOCKED;
+    module->state.post_count = 0;
+    module->state.complete_count = 0;
+    module->epoch_type.access = NONE_EPOCH;
+    module->epoch_type.exposure = NONE_EPOCH;
+    module->lock_count = 0;
+    module->start_group = NULL;
+    module->post_group = NULL;
+    OBJ_CONSTRUCT(&module->outstanding_locks, opal_list_t);
+    module->global_ops_num = 0;
+    module->per_target_ops_nums = calloc(comm_size, sizeof(int));
+
+    win->w_osc_module = &module->super;
+
+    /* sync with everyone */
+
+    ret = module->comm->c_coll->coll_barrier(module->comm,
+                                             module->comm->c_coll->coll_barrier_module);
+
+    if (ret != OMPI_SUCCESS) {
+        goto error;
+    }
+
+    return ret;
+
+ error:
+    if (my_addr) ucp_worker_release_address(mca_osc_ucx_component.ucp_worker, my_addr);
+    if (recv_buf) free(recv_buf);
+    if (module->win_info_array) free(module->win_info_array);
+    if (module->state_info_array) free(module->state_info_array);
+    if (module->disp_units) free(module->disp_units);
+    if (module->comm) ompi_comm_free(&module->comm);
+    if (module->per_target_ops_nums) free(module->per_target_ops_nums);
+    if (module) free(module);
+    return ret;
+}
+
+int ompi_osc_ucx_win_attach(struct ompi_win_t *win, void *base, size_t len) {
+    return OMPI_SUCCESS;
+}
+
+int ompi_osc_ucx_win_detach(struct ompi_win_t *win, const void *base) {
+    return OMPI_SUCCESS;
+}
+
+int ompi_osc_ucx_free(struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+    int i, ret = OMPI_SUCCESS;
+
+    if ((module->epoch_type.access != NONE_EPOCH &&
+         module->epoch_type.access != FENCE_EPOCH) ||
```

Comment:
(now L655)

I think this is the same problem with the first long discussion. I think when the user call Fence with no assert, MPI library cannot decide if epoch is closed or not, and the state should always be FENCE_EPOCH.